### ISSUE TITLE
Backward LSTM results mismatch between TF and ONNXRuntime

### ISSUE THREAD

**Issue by buddhapuneeth** [microsoft/onnxruntime#4187]
**Describe the bug**

Backward LSTM results mismatch between TF and ONNXRuntime. 
in TF runtime (tf.keras), say 1,2,3 are time frames.. output of reverse LSTM will be op3 op2 op1. where as in ONNXRuntime it will be op1 op2 op3. Either ONNX converters need to add a reverseV2 operator after backward LSTM or need to handle on the ONNXRuntime.

**Urgency**
I hack this with change in keras2onnx, but I would like to get confirmation on the expected behavior.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS mojave
- ONNX Runtime installed from (source or binary): pip
- ONNX Runtime version: 1.6
- Python version: 3.6
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source): gcc 11
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**To Reproduce**

[simple_lstm.txt](https://github.com/microsoft/onnxruntime/files/4756547/simple_lstm.txt)
-rename the above file to .h5
-run with some random input on tf.keras backend and on onnxruntime and compare results.
-outputs are just reverse to each other.

Alternatively, you can try this:
code for reverse LSTM:
```
max_features = 10
maxlen = 10
model = Sequential()
model.add(Embedding(max_features, 128, input_length=maxlen))

model.add(LSTM(64, go_backwards=True, return_sequences = True))
model.add(Dense(1, activation='sigmoid'))
model.compile('adam', 'binary_crossentropy', metrics=['accuracy'])


from keras.models import load_model
model.save('simple_rev_lstm.h5')
```

To compare results with other inference engines and ONNXRuntime

```
# pip install onnxruntime
#env TF_KERAS=1
import numpy as np
import onnxruntime
from tensorflow.keras.models import load_model as load_model_tf_keras
np.random.seed(0)
input_data = np.random.randint(10, size=(2, 10))
# with CustomObjectScope({'GlorotUniform': glorot_uniform()}):
sess = onnxruntime.InferenceSession("simple_rev_lstm.onnx")
result = sess.run(["dense_1"], {'embedding_1_input': input_data.astype(np.float32)})
print("ONNX Runtime")
print(np.asarray(result[0]))

model = load_model_tf_keras('simple_rev_lstm.h5')
result = model.predict(input_data)
print("TF Runtime")
print(result)
```
OUTPUT:
```
ONNX Runtime
[[[0.5028585 ]
  [0.51010513]
  [0.5096171 ]
  [0.50567615]
  [0.50046337]
  [0.50202006]
  [0.5010086 ]
  [0.49436894]
  [0.4989641 ]
  [0.50032514]]

 [[0.50235206]
  [0.5045772 ]
  [0.504677  ]
  [0.5032916 ]
  [0.5016299 ]
  [0.50043035]
  [0.4991624 ]
  [0.50064933]
  [0.5028503 ]
  [0.5012636 ]]]

TF Runtime
[[[0.50032514]
  [0.4989641 ]
  [0.49436894]
  [0.5010086 ]
  [0.50202006]
  [0.50046337]
  [0.50567615]
  [0.5096171 ]
  [0.51010513]
  [0.5028585 ]]

 [[0.5012636 ]
  [0.5028503 ]
  [0.50064933]
  [0.4991624 ]
  [0.50043035]
  [0.5016299 ]
  [0.5032916 ]
  [0.504677  ]
  [0.5045773 ]
  [0.50235206]]]
```
Observation:
The values are in reverse to each other on axis 1.
Not sure which is correct behavior.

**Expected behavior**
outputs should be same

**Comment by hariharans29** [comment]
Can you please share `simple_rev_lstm.onnx` ? Probably th converter need to handle this but sharing the converted onnx model will help expedite the investigation. Thanks.

**Comment by buddhapuneeth** [comment]
@hariharans29 You can download from below link:
https://drive.google.com/file/d/1PnvQ5TAR20OWQFfLj8VeVBdpOU6wTMBA/view?usp=sharing

**Comment by buddhapuneeth** [comment]
@hariharans29 any update? Let me know if you need more info'

**Comment by buddhapuneeth** [comment]
@hariharans29 any update?

**Comment by hariharans29** [comment]
Hi - I haven't been able to get to this yet. Will get to this soon.

**Comment by hariharans29** [comment]
Taking a look at this now.

**Comment by hariharans29** [comment]
Hello @buddhapuneeth ,

This is my understanding. The ONNX LSTM spec is unclear on how to return the processed sequence outputs when `direction == reverse`. Per my understanding, it doesn't say that the outputs need to be returned in a reversed fashion. The Keras docs says (https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM), when `go_backwards == true`, "process the input sequence backwards and return the reversed sequence". I think this is more explicit in the intent that the sequence is reversed and returned. (Do you agree ?)

So - 

1) Either the ONNX LSTM spec should be refined to state that the sequence should be reversed for direction == reverse in the next version of the opset and no action is necessary on the converter side
(or)
2) The ONNX LSTM spec should exlicitly state that no reversing is required and the converter must handle this.

**Comment by buddhapuneeth** [comment]
Hi @hariharans29 Yes. I agree.

You rightly mentioned the two options here:
with option#1 All the ONNX runtime implementations like ONNXRuntime, need to adapt it..right?
with option#2 All the convertors need to add reverse_sequence operator after the reverse LSTM. Currently, I made this change for keras2onnx. But there are many other convertors, until someone changes they will have gaps.

Also with the either of the options, there are dependencies with other ONNX libraries. Can you suggest the best approach?

**Comment by hariharans29** [comment]
With option #1 - ONNXRuntime cannot make the change right away. The spec isn't explicit and so this is technically not a bug in ONNXRuntime. In the next version of the op (opset 13), **if** the spec is explicit (if the ONNX community accepts the spec update), ONNXRuntime will make the implementation change (opset 13 implementation only).

With option #2 - I would guess not *all* converters right ? Only for frameworks that expects back the reversed sequence. From what I can see, it is only keras for now. So, I guess this is one for the tf/keras converters. I can't see anything similar (backwards only) in Torch (https://pytorch.org/docs/master/generated/torch.nn.LSTM.html) .

**Comment by buddhapuneeth** [comment]
@hariharans29 thanks for clarifying.  Can I go ahead with my PR in Keras2onnx? or you want to purse option#1. I would like to hear your opinion.

**Comment by hariharans29** [comment]
My suggestion is going ahead with the keras2onnx PR and see if it gets accepted. Pursuing option#1 will take a longer turnaround time (propose to onnx -> have it accepted (may get a push-back) -> wait for onnx release -> wait for ORT implementation change complying onnx release). It seems simpler to go with option#2 given that it seems limited to tf/keras. 

Thanks a lot for bringing this up though.

**Comment by hariharans29** [comment]
Please re-open the issue if you have any new findings.
